### PR TITLE
Feature/controller public dns

### PR DESCRIPTION
**Pull request by ycliuhw**
## Description of change

- make Juju CLI talking with controller using public DNS if possible (because the pod & service address 
 are all internally accessible and pod address might be changed very often) - mainly for k8s controller.
- this PR also enables us to bootstrap in public k8s clusters (like CDK on public clouds, GKE, AKS, EKS, etc)

**Note**: Juju CLI can talk to controllers running on public clusters via `public` service DNS now. But there are some other issues(not directly related with current PR) need to be solved to make the controller fully functioning.

## QA steps

- prepare a public k8s cluster, GKE for example;
- add cloud;

```bash
$ juju add-k8s <k8s-cloud> --local
```
- bootstrap controller;

```bash
$ juju bootstrap <k8s-cloud> <controller-name> --debug
```

- run any cmd to talk with k8s controller, juju status, juju add-model, etc for example;

```bash
$ juju add-model <model-name> <k8s-cloud>
$ juju create-storage-pool mariadb-pv kubernetes storage-class=microk8s-hostpath
$ juju create-storage-pool operator-storage kubernetes storage-class=microk8s-hostpath
$ juju deploy cs:~juju/mariadb-k8s-0  --debug --storage database=500M,mariadb-pv
$ juju deploy cs:~juju/gitlab-k8s-0  --debug  --config juju-external-hostname=127.0.0.1.xip.io
$ juju relate mariadb-k8s gitlab-k8s
$ juju expose gitlab-k8s
```

## Documentation changes

Not ready yet;

## Bug reference

None;

### CODE REVIEW

File: caas/kubernetes/provider/bootstrap.go [hunk 1]
```diff
@@ -98,6 +104,36 @@ func DecideControllerNamespace(controllerName string) string {
 	return "controller-" + controllerName
 }
 
+// ReScheduler executes a function and keep retrying until timeout if the function returns an expected error.
+func ReScheduler(
```
Comment (wallyworld) [inline caas/kubernetes/provider/bootstrap.go]: We already have code for this in the juju/retry repo, eg

```
	err := retry.Call(retry.CallArgs{
		Attempts:    30,
		Delay:       time.Second,
		MaxDelay:    time.Minute, // because 2**29 seconds is beyond reasonable
		BackoffFunc: retry.DoubleDelay,
		Clock:       clock,
		Func: func() error {
			_, err := inst.DeleteSecurityGroup(group)
			if err == nil || isNotFoundError(err) {
				logger.Debugf("deleting security group %q", group.Name)
				return nil
			}
			return errors.Trace(maybeConvertCredentialError(err, ctx))
		},
		IsFatalError: func(err error) bool {
			return common.IsCredentialNotValid(err)
		},
		NotifyFunc: func(err error, attempt int) {
			logger.Debugf("deleting security group %q, attempt %d", group.Name, attempt)
		},
	})
```

File: caas/kubernetes/provider/constants.go [hunk 2]
```diff
@@ -0,0 +1,12 @@
+// Copyright 2019 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package provider
+
+const (
```
Comment (wallyworld) [inline caas/kubernetes/provider/constants.go]: There's consts already in caas/metadata.go

File: caas/kubernetes/provider/k8s.go [hunk 3]
```diff
@@ -816,6 +816,31 @@ func (k *kubernetesClient) DeleteOperator(appName string) (err error) {
...
+	// has different way to report back LB address.
+	// This should cover most of case now.
+	lpAdd := svc.Spec.LoadBalancerIP
+	if lpAdd != "" {
+		return lpAdd
+	}
+
+	ing := svc.Status.LoadBalancer.Ingress
+	if len(ing) == 0 {
+		return lpAdd
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s.go]: This will be ""

File: caas/kubernetes/provider/k8s.go [hunk 4]
```diff
@@ -1613,11 +1634,11 @@ func (k *kubernetesClient) configureService(
...
 		},
 	}
-	return k.ensureService(service)
+	return k.ensureK8sService(service)
 }
 
-// ensureService ensures a service resource.
-func (k *kubernetesClient) ensureService(spec *core.Service) error {
+// ensureK8sService ensures a k8s service resource.
+func (k *kubernetesClient) ensureK8sService(spec *core.Service) error {
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s.go]: why rename?

File: caas/kubernetes/provider/services.go [hunk 5]
```diff
@@ -0,0 +1,16 @@
+// Copyright 2019 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package provider
+
+import (
+	core "k8s.io/api/core/v1"
+)
+
+var preferredControllerServiceTypes = map[string]core.ServiceType{
```
Comment (wallyworld) [inline caas/kubernetes/provider/services.go]: I think this can all go in caas/metadata.go as that's where we have started recording cluster specific preferences like storage etc

File: cmd/juju/common/controller.go [hunk 6]
```diff
@@ -63,9 +66,17 @@ func tryAPI(c *modelcmd.ModelCommandBase) error {
...
+	ctx *cmd.Context,
+	c *modelcmd.ModelCommandBase,
+	isCAASController bool,
+	controllerName,
+	hostedModelName string,
+) (err error) {
 	// TODO(katco): 2016-08-09: lp:1611427
 	attempts := utils.AttemptStrategy{
+		// added total time to try for more stable of when we should give up;
+		Total: bootstrapReadyPollTotalTime,
```
Comment (wallyworld) [inline cmd/juju/common/controller.go]: not sure we need this

File: cmd/modelcmd/base.go [hunk 7]
```diff
@@ -438,6 +439,8 @@ func newAPIConnectionParams(
 	}
 	dialOpts := api.DefaultDialOpts()
 	dialOpts.BakeryClient = bakery
+	// we should not use default timeout - 10 minutes.
+	dialOpts.Timeout = 15 * time.Second
```
Comment (wallyworld) [inline cmd/modelcmd/base.go]: The default should be changed in api.DefaultDialOpts()
10 minutes seems too long but was put there to allow controllers that are upgrading etc to finish.
Maybe a timeout of say 1 minute is better

File: environs/bootstrap/bootstrap.go [hunk 8]
```diff
@@ -144,6 +144,10 @@ type BootstrapParams struct {
 	// CAPrivateKey is the controller's CA certificate private key.
 	CAPrivateKey string
 
+	// ControllerServiceType is the service type to use for a k8s controller.
+	// Optional: only for k8s controller.
+	ControllerServiceType string
```
Comment (wallyworld) [inline environs/bootstrap/bootstrap.go]: We should be able to look this up from the cloud info on the server side.
I don't think there's a need to add a k8s specific attribute here

File: controller/config.go [hunk 9]
```diff
@@ -529,6 +529,15 @@ func (c Config) AutocertDNSName() string {
 	return c.asString(AutocertDNSNameKey)
 }
 
+// SetAutocertDNSName sets controller public DNS.
+func (c Config) SetAutocertDNSName(controllerDNS string) error {
```
Comment (wallyworld) [inline controller/config.go]: This is usually supplied by the user at bootstrap time via the --config argument.
Having bootstrap process overwrite it is problematic.
We could consider adding a cloudServiceDoc for the controller (and also cloudContainerDoc for the pods). This fits better with the data model and allows for when we will model controllers as applications

File: caas/kubernetes/provider/bootstrap.go (hunk 1 above)
Comment (ycliuhw) [reply caas/kubernetes/provider/bootstrap.go]: got it, changed to use this

File: caas/kubernetes/provider/constants.go (hunk 2 above)
Comment (ycliuhw) [reply caas/kubernetes/provider/constants.go]: no, i didn't find it in metadata.go.
it's all using strings currently

File: caas/kubernetes/provider/k8s.go (hunk 3 above)
Comment (ycliuhw) [reply caas/kubernetes/provider/k8s.go]: changed to ""

File: caas/kubernetes/provider/services.go (hunk 5 above)
Comment (ycliuhw) [reply caas/kubernetes/provider/services.go]: moved to metadata.go

File: environs/bootstrap/bootstrap.go (hunk 8 above)
Comment (ycliuhw) [reply environs/bootstrap/bootstrap.go]: as discussed, unfortunately we have to pass through the whole process.
Because k8s cloud type is only known by add-k8s cmd then Juju will lose this knowledge forever.

File: cmd/juju/common/controller.go (hunk 6 above)
Comment (ycliuhw) [reply cmd/juju/common/controller.go]: removed

File: apiserver/admin_test.go [hunk 10]
```diff
@@ -1334,6 +1334,21 @@ func (s *macaroonLoginSuite) TestRemoteUserLoginToControllerLoginAccess(c *gc.C)
...
 	c.Check(result.UserInfo.ControllerAccess, gc.Equals, "login")
 	c.Check(result.UserInfo.ModelAccess, gc.Equals, "")
+	c.Check(result.Servers, gc.DeepEquals,
+		params.FromNetworkHostsPorts(
+			parseHostPortsFromAddress(c, info.Addrs...),
+		),
+	)
+}
+
+func parseHostPortsFromAddress(c *gc.C, addresses ...string) [][]network.HostPort {
```
Comment (wallyworld) [inline apiserver/admin_test.go]: Not sure there's sufficient test coverage here? Maybe it's too hard to create a k8s model and test the service address

File: caas/kubernetes/provider/bootstrap_test.go [hunk 11]
```diff
@@ -54,7 +56,8 @@ func (s *bootstrapSuite) SetUpTest(c *gc.C) {
 	s.controllerUUID = "9bec388c-d264-4cde-8b29-3e675959157a"
 
 	s.controllerCfg = testing.FakeControllerConfig()
-	pcfg, err := podcfg.NewBootstrapControllerPodConfig(s.controllerCfg, controllerName, "bionic")
+	s.controllerCfg[controller.APIPort] = 17070
```
Comment (wallyworld) [inline caas/kubernetes/provider/bootstrap_test.go]: We try and avoid using 17070 in tests. Is there a reason not to stick with 17777 set up in FakeControllerConfig() ?

File: caas/metadata.go [hunk 12]
```diff
@@ -7,11 +7,20 @@ import (
 	"fmt"
 
 	"github.com/juju/collections/set"
+	// core "k8s.io/api/core/v1"
```
Comment (wallyworld) [inline caas/metadata.go]: delete me

File: cmd/juju/caas/add.go [hunk 13]
```diff
@@ -487,6 +488,9 @@ func (c *AddCAASCommand) Run(ctx *cmd.Context) error {
 	if _, ok := newCloud.Config[provider.OperatorStorageKey]; !ok {
 		newCloud.Config[provider.OperatorStorageKey] = operatorStorageName
 	}
+	if _, ok := newCloud.Config[bootstrap.ControllerServiceTypeKey]; !ok {
```
Comment (wallyworld) [inline cmd/juju/caas/add.go]: We need to think about how to handle existing cloud definitions in clouds.yaml. Perhaps default to "LoadBalancer". The out of the box microk8s support will redo any micok8s config and correctly specify cluster ip.

File: cmd/juju/commands/bootstrap_test.go [hunk 14]
```diff
@@ -1804,33 +1810,32 @@ func (s *BootstrapSuite) TestBootstrapPrintClouds(c *gc.C) {
...
-dummy-cloud-with-config                       
-dummy-cloud-with-region-config                
-dummy-cloud-without-regions                   
-many-credentials-no-auth-types                
-
-You will need to have a credential if you want to bootstrap on a cloud, see
-‘juju autoload-credentials’ and ‘juju add-credential’. The first credential
-listed is the default. Add more clouds with ‘juju add-cloud’.
-`[1:])
+	expectedOut := "You can bootstrap on these clouds. See ‘--regions <cloud>’ for all regions.\n" +
```
Comment (wallyworld) [inline cmd/juju/commands/bootstrap_test.go]: Why the change from `` to string concatenation? The original way is better and easier to edit

File: state/cloudservice.go [hunk 15]
```diff
@@ -21,60 +21,78 @@ type CloudService interface {
...
-// Address implements CloudService.
-func (c *cloudService) Addresses() []network.Address {
+// Addresses implements CloudService.
+func (c *CloudService) Addresses() []network.Address {
 	return networkAddresses(c.doc.Addresses)
 }
 
-func (a *Application) cloudService() (*cloudServiceDoc, error) {
-	coll, closer := a.st.db().GetCollection(cloudServicesC)
+func (c *CloudService) cloudService() (cloudServiceDoc, error) {
```
Comment (wallyworld) [inline state/cloudservice.go]: can we continue to return a pointer *cloudServiceDoc

File: state/state.go [hunk 16]
```diff
@@ -1118,6 +1118,47 @@ var (
 	errLocalApplicationExists          = errors.Errorf("application already exists")
 )
 
+type AddCloudServiceArgs struct {
+	// Id will be the application Name if it's a part of application,
+	// and will be controller UUID for k8s a controller(controller does Not have an application).
```
Comment (wallyworld) [inline state/state.go]: not

File: state/state.go [hunk 17]
```diff
@@ -1118,6 +1118,47 @@ var (
...
+	Id         string
+	ProviderId string
+	Addresses  []network.Address
+}
+
+// AddCloudService creates a cloud service.
+func (st *State) AddCloudService(args AddCloudServiceArgs) (_ *CloudService, err error) {
```
Comment (wallyworld) [inline state/state.go]: This needs to be called "SaveCloudService" since the saveServiceOps() below does either an insert or update as needed. AddCloudService() would be a lie sometimes

File: state/state.go [hunk 18]
```diff
@@ -1118,6 +1118,47 @@ var (
...
+	defer errors.DeferredAnnotatef(&err, "cannot add cloud service %q", args.ProviderId)
+
+	doc := cloudServiceDoc{
+		DocID:      args.Id,
+		ProviderId: args.ProviderId,
+		Addresses:  fromNetworkAddresses(args.Addresses, OriginProvider),
+	}
+	svc := newCloudService(st, &doc)
+
+	ops, err := svc.saveServiceOps(doc)
```
Comment (wallyworld) [inline state/state.go]: This needs to be part of a buildTxn loop, to cater for insert vs update races.
See for example (ec *externalControllers) Save() in state/externalcontroller.go

File: state/state.go [hunk 19]
```diff
@@ -1118,6 +1118,47 @@ var (
...
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if err = st.db().RunTransaction(ops); err == nil {
+		if err = svc.Refresh(); err != nil {
```
Comment (wallyworld) [inline state/state.go]: This refresh after save pattern is not one we've used elsewhere have we?
Would prefer to stick to our standard save patterns

File: agent/agentbootstrap/bootstrap.go [hunk 20]
```diff
@@ -391,6 +411,45 @@ func initBootstrapMachine(
...
+
+	env, err := getEnviron(isCAAS, args.ControllerModelConfig, args)
+	if err != nil {
+		return errors.Annotate(err, "getting environ")
+	}
+
+	broker, ok := env.(caas.ServiceGetterSetter)
+	if !ok {
+		// this should never happen.
+		return errors.New("CAAS environ is not a broker")
```
[the remark this answers is not in the document]
Comment (jameinel) [reply agent/agentbootstrap/bootstrap.go]: Would it be better read as:
errors.Errorf("environ %T does not implement ServiceGetterSetter interface")
?

File: agent/agentbootstrap/bootstrap.go [hunk 21]
```diff
@@ -391,6 +411,45 @@ func initBootstrapMachine(
...
+	}
+	svc, err := broker.GetService(k8sprovider.JujuControllerStackName)
```
Comment (jameinel) [inline agent/agentbootstrap/bootstrap.go]: How does this affect things like CDK and microk8s where you may have a DNS name for services, but that DNS only resolves *inside* the K8s (eg, your laptop probably doesn't point to the K8s' DNS)

That is one of the main reason that Juju has historically preferred IP addresses, because DNS is only reliable if it is globally resolvable DNS (and things like MAAS DNS suffer from the same problem.

I know that we do have some support for controllers being advertised at DNS names (it is how we integrate with JAAS). And we do things like negotiate Let's Encrypt to sign our certificates rather than just self-signed.

File: caas/kubernetes/provider/base_test.go [hunk 22]
```diff
@@ -120,7 +120,7 @@ func (s *BaseSuite) setupController(c *gc.C) *gomock.Controller {
 }
 
 func (s *BaseSuite) setupBroker(c *gc.C, ctrl *gomock.Controller, newK8sRestClientFunc provider.NewK8sClientFunc) *gomock.Controller {
-	s.clock = testclock.NewClock(time.Time{})
+	s.clock = testclock.NewClock(time.Now())
```
Comment (jameinel) [inline caas/kubernetes/provider/base_test.go]: Why?
This smells like something is wrong if we need a time that is close to real-time Now. The *point* of Clock is to make sure you don't depend on real time.

File: caas/kubernetes/provider/bootstrap_test.go (hunk 11 above)
Comment (jameinel) [reply caas/kubernetes/provider/bootstrap_test.go]: shouldn't this be a dynamic port? or we know that this is never listened on or talked to so it is arbitrary?
I agree you want to avoid 17070 and have things "accidentally" work, and ignore the config

File: caas/kubernetes/provider/k8s.go [hunk 23]
```diff
@@ -817,6 +817,31 @@ func (k *kubernetesClient) DeleteOperator(appName string) (err error) {
 	return errors.Trace(k.deleteDeployment(operatorName))
 }
 
+func getLoadBalancerAddress(svc *core.Service) string {
+	// reason we have this is because different cloud provider
+	// has different way to report back LB address.
+	// This should cover most of case now.
```
Comment (jameinel) [inline caas/kubernetes/provider/k8s.go]: different cloud providers have a different way to report back the Load Balancer address.
This covers the cases we know about so far.

File: cmd/juju/commands/bootstrap.go [hunk 24]
```diff
@@ -750,36 +757,64 @@ See `[1:] + "`juju kill-controller`" + `.`)
...
-			AgentVersion:           agentVersion.String(),
-			CurrentHostPorts:       [][]network.HostPort{network.AddressesWithPort(addrs, config.controller.APIPort())},
-			PublicDNSName:          newStringIfNonEmpty(config.controller.AutocertDNSName()),
-			MachineCount:           newInt(1),
-			ControllerMachineCount: newInt(1),
-		}); err != nil {
-		return errors.Annotate(err, "saving bootstrap endpoint address")
+
+	retryCallArgs := retry.CallArgs{
+		Attempts:    20,
```
Comment (jameinel) [inline cmd/juju/commands/bootstrap.go]: we already have some retry logic for IAAS controllers because the time it takes to boot a machine can vary wildly, especially with MAAS. And we already have Config to allow users to set those things.
Setting a max retry timeout of 1minute seems very aggressive (our default timeout is 10min)
Is there a reason we need yet-another retry loop here that doesn't use our existing timeout configuration?

File: state/state.go (hunk 19 above)
Comment (jameinel) [reply state/state.go]: we should be using st.db().Run(buildTxn) if we can, instead of RunTransaction, as it allows you to revalidate your assertions and give better error messages along with retrying if the db has changed vs memory.

File: state/state_test.go [hunk 25]
```diff
@@ -1510,6 +1510,27 @@ func (s *StateSuite) TestAllRelations(c *gc.C) {
...
+	c.Assert(state.LocalID(s.State, svc.Id()), gc.Equals, "cloud-svc-ID")
+	c.Assert(svc.ProviderId(), gc.Equals, "dfsds-sdfsd-sdfds")
+	c.Assert(svc.Addresses(), gc.DeepEquals, network.NewAddresses("1.1.1.1"))
+
+	getResult, err := s.State.CloudService("cloud-svc-ID")
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(state.LocalID(s.State, getResult.Id()), gc.Equals, "cloud-svc-ID")
+	c.Assert(getResult.ProviderId(), gc.Equals, "dfsds-sdfsd-sdfds")
+	c.Assert(getResult.Addresses(), gc.DeepEquals, network.NewAddresses("1.1.1.1"))
+}
```
Comment (jameinel) [inline state/state_test.go]: Ideally we would add tests that use the transaction Hooks to ensure we are properly doing assertions against the database. (the hooks let you make a change right as the 'transaction' is starting, so that you can see that even though in-memory we think the content is Foo, we properly assert against the database, and notice that the content was changed by someone else to Bar.)

File: cmd/juju/commands/bootstrap.go (hunk 24 above)
Comment (ycliuhw) [reply cmd/juju/commands/bootstrap.go]: ah, I have moved this retry into `provider.Bootstrap()`. I probably should remove this retry here.
The reason why we  need the 2nd retry strategy for k8s controller is we need to ensure LB provisioning has been finished (LB address is ready to use) and we need the LB address in bootstrap-state cmd or the cmd will fail.
So we wait and fail in juju CLI if we can't get the address to avoid failing in bootstrap-state in Jujud.

File: agent/agentbootstrap/bootstrap.go [hunk 26]
```diff
@@ -391,6 +411,45 @@ func initBootstrapMachine(
...
+		return errors.Trace(err)
+	}
+	if len(svc.Addresses) == 0 {
+		// this should never happen because we have already checked in k8s controller bootstrap stacker.
+		return errors.NotProvisionedf("k8s controller service %q address", svc.Id)
+	}
+	svcId := args.ControllerConfig.ControllerUUID()
+	logger.Infof("creating cloud service for k8s controller %q", svcId)
+	cloudSvc, err := st.AddCloudService(state.AddCloudServiceArgs{
+		Id:         svcId,
```
[the remark this answers is not in the document]
Comment (ycliuhw) [reply agent/agentbootstrap/bootstrap.go]: yeah, the arg.Id is wrapped with `applicationGlobalKey` in `state.AddCloudService` now.

File: apiserver/admin_test.go (hunk 10 above)
Comment (ycliuhw) [reply apiserver/admin_test.go]: i was trying to add test for CAAS controller model, 
but i got some difficulty to setup a CAAS controller model because it's using `jujutesting.JujuConnSuite`..

File: caas/kubernetes/provider/bootstrap_test.go (hunk 11 above)
Comment (ycliuhw) [reply caas/kubernetes/provider/bootstrap_test.go]: right, fixed!
now the expected spec is read from controllerCfg.
thanks

File: apiserver/admin_test.go (hunk 10 above)
Comment (wallyworld) [reply apiserver/admin_test.go]: There's a pattern we use for that, search for 
// Set up a CAAS model to replace the IAAS one.

**Comment by wallyworld** [comment]
> I think I saw that the default behavior is to use an IP if the service describes it, and then to fall back to a hostname if no IP is available. Which sounds likely to be the right behavior for most locations. But I'd like to confirm the intent.

We can't use hostname because the name resolution behaviour seems inconsistent across clusters, and on microk8s at the moment, doesn't seem to comply with the k8s docs. There's an internal cluster level endpoint we are looking to use in the future but for now the service external IP is all we have that will work across clusters

File: agent/agentbootstrap/bootstrap.go (hunk 21 above)
Comment (wallyworld) [reply agent/agentbootstrap/bootstrap.go]: The idea is we send the public service IP address to JUju clients (which is what this code is getting).
Internally we want to use something else, and ostensibly there's an internal cluster DNS and a naming scheme https://kubernetes.io/docs/concepts/services-networking/dns-pod-service. But experiments have shown that not everything behaves as documented at the moment.

File: state/state_test.go (hunk 25 above)
Comment (ycliuhw) [reply state/state_test.go]: very handy, added 2 more tests using hooks for negative cases; thanks

File: agent/agentbootstrap/bootstrap.go [hunk 27]
```diff
@@ -289,6 +283,32 @@ func ensureHostedModel(
 	return nil
 }
 
+func getEnviron(isCAAS bool, modelConfig *config.Config, args InitializeStateParams) (env environs.BootstrapEnviron, err error) {
```
[the remark this answers is not in the document]
Comment (ycliuhw) [reply agent/agentbootstrap/bootstrap.go]: you are right, changed,

File: state/cloudservice.go (hunk 15 above)
Comment (ycliuhw) [reply state/cloudservice.go]: changed

File: state/state.go (hunk 18 above)
Comment (ycliuhw) [reply state/state.go]: yup, changed

File: caas/kubernetes/provider/base_test.go (hunk 22 above)
Comment (ycliuhw) [reply caas/kubernetes/provider/base_test.go]: ah just changed to Now without thinking because I saw somewhere else uses `time.Now()`, 
yeah , u r right, it could be any time.

File: apiserver/admin_test.go (hunk 10 above)
Comment (ycliuhw) [reply apiserver/admin_test.go]: I mean to setup a CAAS `controller` model which requires refactor a lot on `jujutesting.JujuConnSuite` itself.

File: cmd/juju/commands/bootstrap_test.go (hunk 14 above)
Comment (ycliuhw) [reply cmd/juju/commands/bootstrap_test.go]: the original one has lots of trailing spaces on most of lines.
tests would be failed due to those spaces are automatically removed by linter..
it's quit annoying that I have to disable linter whenever i have to `ctrl+s` the file.

File: cmd/juju/caas/add.go (hunk 13 above)
Comment (ycliuhw) [reply cmd/juju/caas/add.go]: yeah, the result uses LB now by default unless it can find related definition for this cloud.

**Comment by ycliuhw** [comment]
!!build!!

**Comment by ycliuhw** [comment]
$$merge$$

**Comment by ycliuhw** [comment]
[introduce public DNS for Juju controller](https://trello.com/c/0P1rLr44/681-introduce-public-dns-for-juju-controller)
